### PR TITLE
Updates to finetuning functionality for the multicenter study

### PR DESCRIPTION
**Pull request by delaneysmith22**
This PR contains: 
1. A re-vamp/re-design of a single trial of the "cucumber" method
- external_validation.py now runs a single trial, not an entire experiment (5 trials), for which you can manually specify the seed for splitting the dataset into folds (facilitates reproducibility).
2. The option to upsample from the minority class by generating more than 1 m-mode from each clip during pre-processing
3. Functionality for mid-training subgroup analysis (stratifying metrics by center, probe type, sex, etc.)
4. Updated functionality for plotting summarized and stratified finetuning results (in finetune_results_utils.py)

### CODE REVIEW

File: config.yml [hunk 1]
```diff
@@ -97,7 +97,7 @@ TRAIN:
       LR_DECAY_VAL: 0.05642  # Learning rate decay
       LR_DECAY_EPOCH: 15 # Epoch to start lr decay
       DROPOUT: 0.2151 # Dropout probability
-      BLOCKS_FROZEN: 30 # Number of blocks to freeze - MUST be 0 or positive
+      LAYERS_FROZEN: 220 #118 #161 # Number of blocks to freeze - MUST be 0 or positive
```
Comment (bvanberl) [inline config.yml:100]: Could we change this (and references in the code) to refer to layers frozen and not blocks?

File: src/data/download_videos.py [hunk 2]
```diff
@@ -76,20 +77,23 @@ def download(df, sliding, fr_rows, video_out_root_folder=cfg['PATHS']['UNMASKED_
...
+#
+# # Establish connection to database
+# cnx = mysql.connector.connect(user=USERNAME, password=PASSWORD,
+#                               host=HOST,
+#                               database=DATABASE)
+#
+# # Query Database for sliding and no_sliding labeled data but excluding significant probe movement and B lines
+# # and pleural effusion or consolidation
+# sliding_df = get_clips_from_db(cfg['PARAMS']['DB_TABLE'], sliding=True)
+# no_sliding_df = get_clips_from_db(cfg['PARAMS']['DB_TABLE'], sliding=-False)
```
Comment (bvanberl) [inline src/data/download_videos.py:93]: Maybe we should uncomment this code and provide an option to use predefined sliding and no sliding CSVs

File: src/data/download_videos.py [hunk 3]
```diff
@@ -99,11 +103,11 @@ def download(df, sliding, fr_rows, video_out_root_folder=cfg['PATHS']['UNMASKED_
...
-no_sliding_extra_df = pd.read_sql('''SELECT * FROM clips_chile WHERE (pleural_line_findings='absent_lung_sliding'
-                                      OR pleural_line_findings='thickened|absent_lung_sliding') AND
-                               (quality NOT LIKE '%significant_probe_movement%' OR quality is null) AND
-                               (a_or_b_lines='a_lines' OR a_or_b_lines='non_a_non_b' OR a_or_b_lines is null) AND
-                               (pleural_effusion is null) AND (consolidation is null);''', cnx)
+# no_sliding_extra_df = pd.read_sql('''SELECT * FROM clips_ottawa WHERE (pleural_line_findings='absent_lung_sliding'
+#                                       OR pleural_line_findings='thickened|absent_lung_sliding') AND
+#                                (quality NOT LIKE '%significant_probe_movement%' OR quality is null) AND
+#                                (a_or_b_lines='a_lines' OR a_or_b_lines='non_a_non_b' OR a_or_b_lines is null) AND
+#                                (pleural_effusion is null) AND (consolidation is null);''', cnx)
```
Comment (bvanberl) [inline src/data/download_videos.py:110]: The above comment applies here too

File: src/data/save_m_modes.py [hunk 4]
```diff
@@ -45,19 +48,81 @@ def get_clip_length(video_path):
...
+    #         # Get clip length.
+    #         l = get_clip_length(os.path.join(clip_class_dir, clip))
+    #         if l < 3.0:
+    #             short_clips.append(os.path.join(clip_class_dir, clip))
+    # # Show the filenames of all clips less than the mini-clip window specified in the config file.
+    # for clip in short_clips:
+    #     print(clip)
+    # print(len(short_clips))
+    #
+    # save_m_modes(npz_dir, mmode_dir)
```
Comment (bvanberl) [inline src/data/save_m_modes.py:128]: I think we can delete these comments

File: src/data/to_npz.py [hunk 5]
```diff
@@ -85,7 +85,8 @@ def miniclip_to_mmode(clip, bounding_box, height_width):
     '''
```
Comment (bvanberl) [inline src/data/to_npz.py:85]: Documentation for `method` missing

File: src/external_validation.py [hunk 6]
```diff
@@ -92,428 +95,566 @@ def check_performance(df):
...
+            extra_linear_example = clips_available.iloc[i % clips_available.shape[0]]
+            extra_linear_example['filename'] = os.path.splitext(extra_linear_example['filename'])[0] + f'_{extra_idx}.npz'
+            dataset = dataset.append(extra_linear_example)
+
+    return dataset
+
+
+def reinitialize_layer(model, initializer, layer_name):
+    '''
+    Re-initializes the weights of a specific layer of the model froma  desired distirbution.
```
Comment (bvanberl) [inline src/external_validation.py:345]: "froma" typo

File: src/external_validation.py [hunk 7]
```diff
@@ -92,428 +95,566 @@ def check_performance(df):
...
+    '''
+    Fine-tunes the model on a training set.
+    :param original_model_path: Path to SavedModel constituting the original weights
+    :param base_folder: The folder in which logs, models, and train/val splits will be saved
+    :param full_train_df: The training set (before splitting off a validation set)
+    :param mmode_preprocessor: a MModePreprocessor object for preparing batches
+    :param hparams: Model hyperparameters
+    :param upsample: If set to True, upsamples the minority class in the training set according to the configured ratio
+    :param seed: Random seed for fold sampling
+    :param AB_scheme: If True, fine-tunes using the approach from our AB manuscript (10 epochs at LR, n epochs at LR/10)
```
Comment (bvanberl) [inline src/external_validation.py:366]: Maybe we should link the AB paper or repo here

File: src/external_validation.py [hunk 8]
```diff
@@ -92,428 +95,566 @@ def check_performance(df):
...
+
+    # Save and return best model's weights
+    model_path = os.path.join(base_folder, f'model')
+    model.save(model_path)
+
+    return model
+
+
+def TAAFT(clip_df, n_folds, experiment_path=None, seed=None, hparams=None, stratify_by=None):
+          #by_center=False, by_probe=False):
```
Comment (bvanberl) [inline src/external_validation.py:508]: For now let's remove the above comments

File: src/models/models.py [hunk 9]
```diff
@@ -767,10 +769,11 @@ def efficientnet(model_config, input_shape, metrics, class_counts):
     l2_reg = model_config['L2_REG']
     fc0_nodes = model_config['FC0_NODES']
 
-    freeze_cutoff = -1 if model_config['BLOCKS_FROZEN'] == 0 else model_config['BLOCKS_FROZEN']
+    freeze_cutoff = -1 if model_config['LAYERS_FROZEN'] == 0 else model_config['LAYERS_FROZEN']
 
     X_input = Input(input_shape, name='input')
-    base_model = EfficientNet(include_top=False, weights='imagenet', input_shape=input_shape, input_tensor=X_input)
+    X_gray = ToGrayscale(name='to_grayscale')(X_input)
```
Comment (bvanberl) [inline src/models/models.py:775]: Let's get rid of ToGrayScale in this model for now -- I just added it to test its effect on the WB device

**Review by bvanberl — APPROVED** [review]

File: config.yml (hunk 1 above)
Comment (delaneysmith22) [reply config.yml:100]: I didn't find any reference to blocks frozen in external_validation.py or train.py.

**Pull request merged**